sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,1078 @@
+  kernel_handler LocalKernelHandler;
+  LocalKernelHandler.__init_specialization_constants_buffer(specialization_constants_buffer);
+  // for simplicity we could have just used
+  // kernel_handler LocalKernelHandler = { args_for_kernel_handler_init };
+  // here, but we assume that kernel_handler might be used for more than just
+  // accessing specialization constants and therefore there could be other
+  // initialization parameters which also could be conditional
+  // Even now we don't need to always initialize the kernel_handler object
+  // Re-used body of "sycl_kernel" function:
+  {
+     LocalLambdaClone(LocalKernelHandler);
+   }
+}
+```
+
+As mentioned above, creation of `specialization_constants_buffer` kernel
+argument and initialization of `LocalKernelHandler` object with it only happens
+if we are not targeting SPIR-V, i.e. when we compile code for a target without
+native support for specialization constants.
+
+If that new argument was added, it is communicated to DPC++ through regular
+integration header mechanism, i.e. it is added as new entry to
+`kernel_signatures` structure there with parameter kind set to a new
+enumeration value `kernel_param_kind_t::kind_specialization_constants_buffer`.
+
+#### `__builtin_sycl_unique_id`
+
+This built-in is used to generate unique identifiers for specialization
+constants, which are used in communication between the compiler and the runtime.
+
+`__builtin_sycl_unique_id` is defined as follows: it accepts a variable and returns
+a C-string (`const char *`), which:
+- if the input variable has external linkage, the string must be the same in all
+  translation units that pass this same variable to the built-in.
+- if the input variable has internal linkage, the string must be unique across
+  all translation units.
+- return string must be the same if the built-in was called twice for the same
+  variable within a single translation unit (regardless of its linkage type).
+
+#### Integration footer generation
+
+Note: we could have used `__builtin_sycl_unique_id` directly in DPC++ Headers, but
+this would break compilation of those with a third-party C++ 17-compatible
+compiler, which is unaware of this built-in function. Therefore, the compiler
+generates a header file, which includes _the result_ of calling
+`__builtin_sycl_unique_id` function and it is included into the user's program. By
+doing so we can still use this non-standard built-in function and preserve
+support for third-party host compilers.
+
+However, as noted above, we can't use regular integration header here, because
+in general case, `specialization_id` variables can't be forward-declared.
+Therefore, we are using _integration footer_ approach, i.e. we generate a header
+file which must be included at the end of a translation unit.
+
+For the following code snippet:
+```
+struct A {
+  float a, b;
+};
+
+constexpr specialization_id<int> id_int;
+struct Wraper {
+public:
+  static constexpr specialization_id<A> id_A;
```

Comment:
We should also mention wrappers which are being introduced in the patch which does "integration footer integration".

Reply:
They are mentioned in the section `Ambiguous references to specialization_id`